File: public/dsa/arrays/largest_contiguous_sum/kadane/largest_contiguous_sum.cpp
#include <bits/stdc++.h>
using namespace std;

int largestContiguousSum(vector<int> &arr) {
    int max_so_far = arr[0];
    int max_ending_here = arr[0];

    for (int i = 1; i < arr.size(); i++) {
        max_ending_here = max(arr[i], max_ending_here + arr[i]);
        max_so_far = max(max_so_far, max_ending_here);
    }
    return max_so_far;
}

// tests-start
int main() {
    vector<int> arr = {4, -6, 2, 5};
    cout << largestContiguousSum(arr) << endl;
    return 0;
}
// tests-end

/*output
7
*/

File: public/dsa/arrays/largest_contiguous_sum/kadane/largest_contiguous_sum.py


from typing import List


class Solution:
    def largestContiguousSum(self, arr: List[int]) -> int:
        max_so_far = arr[0]
        max_ending_here = arr[0]

        for i in range(1, len(arr)):
            max_ending_here = max(arr[i], max_ending_here + arr[i])
            max_so_far = max(max_so_far, max_ending_here)

        return max_so_far


# tests-start
arr = [4, -6, 2, 5]
sol = Solution()
print(sol.largestContiguousSum(arr))
# tests-end

'''output
7
'''
